### DISCUSSION TITLE
Update hitarea after resize

### DISCUSSION THREAD

**Discussion by lionettiluca** [pixijs/pixijs#7282] [Q&A]
Hi guys,

I've a problem when I try to catch event on fullscreen hitarea of container after window resize.
In my case, I need to mantain the central stage (in the example initial the green square) always in the center of screen, also after resize) and add other squares around by click.
The stage is ok, the container resize is enought, but after resize the coordinates of the click are wrong.

I've read about moving every single element after resize but in my case could be a problem for the number of elements.

Any idea or suggestion about?
Thanks in advance 🙏

- **`pixi.js` version**: 5.3.7
- **Browser & Version**: Chrome 88
- **Running Example**: https://jsfiddle.net/LkLio/1cykrpds/16/

**Comment by ivanpopelyshev** [comment]
There's a shortcut for hitArea/filterArea stuff:

```
container.hitArea = app.screen;
```

**Comment by lionettiluca** [comment]
Thanks @ivanpopelyshev for reply, but sadly It doesn't solve the problem after resize.
After resize the hitarea seems translated respect the container and, at click, element result translated respect mouse
I think it could be a position problem but I'm not sure

**Comment by ShukantPal** [comment]
Your drawing code is too complicated, and that's what's causing your confusion. To start - the hit-area is defined _in the local space of the target object_. That implies it will move with it.

There are three things that depend on the dimensions of the window:
* `container.position` (x, y)
* `container.pivot`
* `center.geometry.graphicsData[0].shape` (i.e. the `drawRect` call)

Your code only updates the first dependency. As such, the container "appears" to shift the center rectangle in-step with the screen. Except it doesn't. Let me explain.

### Shifting of the origin

The local space of any DisplayObject, here the `container`, has a origin = (0, 0) and a pivot (px, py). In the parent's space, here the `stage` or `world` or the canvas view, the pivot gets mapped to the `position` (x, y). Assuming no scaling/rotation, where does the origin get mapped to in the parent?

(x - px, y - py)

```js
// Initialization
container.x = app.screen.width / 2
container.y = app.screen.height / 2
container.pivot.x = window.innerWidth / 2
container.pivot.y = window.innerHeight / 2
```

You initialize the container so that the origin is equal to (0, 0) in the canvas view. (i.e. app.screen.width / 2 - window.innerWidth / 2 = 0, app.screen.height / 2 - window.innerHeight / 2 = 0)

When you resize, however, you aren't updating the pivot. So what happens: 

```js
container.x = app.screen.width / 2
container.y = app.screen.height / 2
```

The origin of the container now becomes (newWidth / 2 - oldWidth / 2, newHeight / 2 - oldHeight / 2). You never moved the hitArea.

### Implications

```js
container.on('pointerdown', (event) => {
	let rect = new PIXI.Graphics()
		.beginFill(0xff0000)
  	.drawRect(event.data.global.x, event.data.global.y, 100, 100);
   
	container.addChild(rect)
})
```

Since you shifted the origin of the container, anything you add as a child in the container will also be shifted by that amount. Since you are using the "global" coordinates of the pointer (which remain same after shifting), adding a graphics with those coordinates will also shift.

### Fix

You need to update your `hitArea` on resize() as well.

#### 1. Fix your origin shifting

```js
// resize()
container.x = app.screen.width / 2
container.y = app.screen.height / 2
container.pivot.x = window.innerWidth / 2 // or app.screen.width / 2
container.pivot.y = window.innerHeight / 2 // or app.screen.height / 2

// Also redraw the center rect because it needs to redraw in its local space, without shifting its own origin. This
// means the hit-area won't have to be readjusted
center.clear()
        .beginFill(0x00FF00)
	.drawRect(window.innerWidth / 2 - 50, window.innerHeight / 2 - 50, 100, 100)
```


#### 2. Transform global pointer coordinates into local space

```js
container.on('pointerdown', (event) => {
        // Move into container's local space since we are drawing in an equivalent space
        let loc = event.data.getLocalPosition(container);
	let rect = new PIXI.Graphics()
		.beginFill(0xff0000)
  	        .drawRect(loc.x, loc.y, 100, 100);
   
	container.addChild(rect)
})
```

**NOTE: The hitArea will remain shifted with this solution. This doesn't address the shifting of container, that's why**

#### 3. Don't use the 3 degrees of freedom together - position, pivot, and drawRect

Instead of tinkering with position + pivot and then a shifted drawRect, make your drawRect centered at the DisplayObject's origin and then set the position only on the `center` instead of `container`.

```js
center.beginFill(0xff0000)
  	   .drawRect(-50, -50, 100, 100);
// Update this on resize() instead
center.position.set(window.innerWidth / 2, window.innerHeight / 2);

// Never move container!
container.hitArea = new PIXI.Rectangle(0, 0, window.innerWidth, window.innerHeight / 2);
```

**Reply by lionettiluca** [reply]
Wow @SukantPal thank you so much for your super explanation! I've solved all problems 🙏

The most interest topic is about container coordinates and its pivot. 
The only difference from your example is the reset of center element position at resize.
I've set container.pivot (0,0) and all element based on It. With this, after resize, All elements move with stage and the center of stage result always at the center of window!

I hope it is a simpler solution 😅

If someone may need it, I've updated demo: https://jsfiddle.net/LkLio/1cykrpds/49/
